### DISCUSSION TITLE
Photos/Lightroom Classic work"flow" - long & ranty (and HDR)

### DISCUSSION THREAD

**Discussion by cfc62** [RhetTbull/osxphotos#1617] [Ideas]
I apologize in advance as this won't be a clear cut, crisp post. It's borne out of frustration; things that should work don't, things that have worked don't, and change is a constant. 

I used the "ideas" tag because I do think something can come out of these struggles, necessity is the mother of all invention and all that. 

**Background:**

I take photos. Lots of photos. I use a variety of gear, not just iPhones. I shoot RAW. 

In general and overall (exceptions exist), I like the Apple Photos ecosystem. I don't participate fully, as I shun iCloud Photos other than a small shared album for extended family members to contribute to. 

Specifically, I like being able to take photos, easily import them into Photos, lightly edit them, and then have generally pleasing photos to view via slideshows either on my Mac or more frequently, on my TV via AppleTV and the Sharing app. 

Don't get me wrong, there have been a ton of problems over the years with the Sharing app, but over the last 2-3 years it's been behaving OK. Or at least I know how to make it work most of the time (lots of reboots are involved after changes are made to the library).

There's a flow, and it's... OK. But only OK. Specifically, editing is basic. Sure, it will do RAW and HDR, but in pretty limited ways. It's no longer enough for me. 

Enter Adobe's Lightroom, specifically Lightroom Classic (again shunning their cloud product). I'll be honest, I do not like Adobe. At all. At the same time, this product does everything in photo editing anyone would want. It's dominant, and will remain that way for the foreseeable future. There's also a whole ecosystem around Adobe's photography products that just doesn't exist for Apple Photos. Need help? You can get it, there are probably 20-30 Youtube tutorials on how to do exactly what you're looking for. 

**Problem:**

The Adobe products do things well, except play well with Apple Photos. 

Both Lightroom Classic (LrC) and Photos want to do many of the same things; they both want to be the Digital Asset Manager (DAM) for your photos. I also think each crowd of users is suspicious of the other product, sometimes with validity, sometimes you can tell people are saying things they think are true but are not. 

In many ways, LrC is overbuilt and has too many features for most people. It's meant for pro-level photographers with clients, and you can see this in the interface and functions they have. They've tried to create a new product called Lightroom which is cloud-based, but this is a) too simple b) lack important functionality and c) cloud. 

To be frank, I want to use LrC for one thing, and one thing only - editing photos, one at a time. I do not want or need LrC to be my DAM. But I do need the editing power that it has. 

Hope:

I'd like to be able to select a photo in Photos, right-click on it and select "Edit in Lightroom Classic" and go edit, then when done automatically have the edited result stuffed back into the library with metadata intact. 

Reality:

To be clear, this is possible today, but with issues. First of all, adding LrC as an app that can edit photos isn't permanent. It tends to disappear from the contextual menu. Secondly, you're not getting the original file, much less the RAW file, but rather a .jpeg file which incorporates all of the edits you may have made to the photo. Clearly that's a non-starter. 

So, you might correctly say, export the unmodified original file. (Note, I'm fully OK with losing whatever edits I might have made, and you'll see why in a bit.)

This entails just a tad more work to get the file into LrC. And it's important to know that LrC (like Photos) *requires* photos to be imported. Again, it insists on being your DAM so there's another step involved here before you get to the sweet sweet editing tools in LrC. Even if you select "Edit in Lightroom" on a photo you still need to import it in LrC before editing. But at least this way we can use a RAW or DNG if we have it. 

OK, so you edit the file. Sweet. Now it's time to export the file. Except the file format that Adobe will write are not necessarily the file formats Apple will accept. There are three, no four parts to this. One is whether or note Photos will even accept a file. Then there's the question if Photos will render the file correctly. Then there are formats that Photos will accept, will render correctly, but won't show up (or will show up botched) in slideshow mode, and the same is true for ATV sharing. 

Here, it at least _looks_ like AVIF may be the file format of choice going forward, but support within the Apple ecosystem is spotty. Most AVIF formats import to Photos, but some don't. More frustratingly, none of the AVIF formats make it through slideshows at all, and many are not displayed correctly in Preview. 

Much like how there was an industry fight over HD and 4K video formats, I suspect there's something similar going on behind the scenes with the newer formats. 

In any case, once you've found a file format that (somewhat) works, importing the result back into Photos is another drag and drop or import via menu. Nicely enough as long as the metadata is preserved through the process (easy enough, you'd have to try to delete it), the new edited photo will be right next to the original. I like this result, and this is why I said earlier that I didn't mind losing previous edits, because the "original" in this case can will have whatever edits I applied to it earlier. I haven't lost anything unless I want to delete the original if I think the edits in LrC are the best I can do. 

Even as I write this out I can see that perhaps this just isn't going to work. But perhaps it could; Adobe does have an API that I need to investigate, and they allow plugins. 

I have a highly related rant on HDR coming up next...

**Comment by RhetTbull** [comment]
You might be able to automate some of this with AppleScript. Certainly the export original from Photos is easily scriptable. I don't use LR Classic (or any Adobe product if I can help it) but I don't think it's scriptable based on a quick search. However, you could use accessibility scripting to send keystrokes to LR to import the asset and then re-import back to Photos. If you know Python, my [photoscript](https://github.com/RhetTbull/PhotoScript) package provides a python wrapper around the Photos AppleScript interface. If you don't, then ChatGPT is actually pretty good at generating AppleScript so it might be able to help. 

I have a couple related ideas / issues but not directly what you're trying to do:

#909 to add a new command to transform a photo -- that is, export it, do some transformation, then re-import it and replace the original, re-applying all metadata
#1233 - bi directional sync with Photos. That is, do an export, then osxphotos monitors any changes to the exported image and then applies the edits to the original in Photos.

Not sure when I'll get to these but I think they'd be helpful for workflows such as yours.

**Reply by cfc62** [reply]
Thanks - I'll look at this. I appreciate you taking the time to go through my word salad. 

I'm still going to rant about HDR though. 

And I downloaded Sequoia yesterday, I may rant about Photos now too. But I need more time to see how the changes work out. Initial impressions are decidedly mixed. 

As for automating LrC, they don't have an API so to speak but really an SDK. They use Lua as the language, but you're correct in that they only allow extensibility in certain areas, and neither importing or editing are included.

**Comment by cfc62** [comment]
So, let's talk about HDR for a bit. 

_Do not underestimate HDR's impact on photography._ For sure, lots of people have seen bad HDR and written it off. But let's be clear, good HDR on a good HDR display is simply _transformational_. 

No, it doesn't and can't translate to printed material, or to regular SDR displays. But the vast majority of people no longer print; photos are viewed on screens. And so many screens these days are already HDR. 

But I've been shocked (positively) at how many times an image I thought I botched in shooting actually winds up being very good or even spectacular when viewed in HDR. Many times I'd have the thought "that's what it really looked like!" when converting a RAW into HDR in LrC. 

But it's more than just RAW that benefits. I'm finding that almost any photo gets a great boost from HDR. Even 20 year old 4MP JPG files can (and do) pop. I wouldn't have thought that possible - but there's more dynamic range that can be unlocked. It's not required to have a 10, 12 or 14 bit RAW file to see benefits, although of course that's where the really cool stuff will happen. 

Needless to say, I've been really stoked about this. But the software support is spotty and inconsistent at best. And there are caveats all over the place - probably the biggest one is that displaying HDR images on SDR software or displays gives poor results. 

At the same time, it's frustrating to see how well HDR video has been accepted in software and by providers of all types. 4K HDR on YouTube is an example; it works great on my Mac and ATVs. File formats are set and accepted, everyone is moving forward.

As of Sonoma, Apple's support for HDR was pretty bad. Preview didn't support it, and you could only see HDR in parts of Photos. There's some support in Safari, but it's not clear to me how complete it is. I don't see any support for HDR on the Apple TV for photos. This is odd as Apple's been capturing HDR on iPhones for years now. There's a huge amount of photos out there that can benefit from a complete HDR flow.

I'm running Sequoia, and while there's _apparently_ more support for HDR overall, it's not immediately apparent. I have more work to do to investigate for sure. In Photos there's a setting to "Display HDR in full" or something like that, but honestly I'd say the support for HDR looks about the same as Sonoma. Do I see images in HDR? Yes, but it appears that HDR display is reserved for only full screen or editing. I suspect Apple thinks showing HDR and SDR images at the same time is going to be too jarring. 

Apple frustratingly doesn't allow for RAW editing into HDR. I had hoped this would be a part of Photos in Sequoia. Nope.

LrC doesn't do much better, but it does support editing in HDR, which is key. But at the same time many windows just don't show images in HDR. Basically only the Develop window supports HDR. You can't even go to full screen (F) and see HDR. 

But what LrC critically does is allow for any photo to be upgraded into HDR. Yes, you need to click on the HDR button and then make adjustments to each photo, but it lets you do it. Photos does not. Perhaps Apple thinks old JPEGs can't become good HDR content; I can't blame them for thinking that, as that's what I thought too. But not allowing HDR edits for (non-iPhone) RAW files is a miss, and a big one at that. 

Are 20 year old JPEGs going to be magically full HDR? No, of course not. But will they look much better and more realistic? You better believe it. 

In LrC, the process is roughly:

- Open the photo in Develop (d) - _again, this is really the only place where you'll see the photo in HDR_
- Hit the HDR button
- Say "wow that looks a lot better"
- Tweak as desired
- Export the file

This last step is fraught with pitfalls. Oh, not the exporting itself, that's fine. It's whether or not Apple accepts the file as HDR and shows it properly, and does so throughout the ecosystem. As of now, that's absolutely not the case, so my ardor for going through my library and converting to HDR has cooled substantially. But I was hoping Sequoia would bring some major steps forward, and I'm not seeing that just yet. I have more testing to do, to be clear. 

Even on Sonoma things that worked a couple of months ago don't work now in the 14.6 beta. Bug? Intentional regression? Unclear, but frustrating. 

So, I'm positively stoked to upgrade to HDR, but I can't do it at the moment. LrC isn't the issue, Apple is. Yes Adobe can and likely will extend HDR support throughout LrC, but the lack of 100% coverage is not a showstopper at the moment. 

If I sound frustrated, it's because I am. There's real value and joy to be unlocked here. I suspect there are technical and technical-political battles going on around file formats and user experience and all that. Apple gets very picky about what file formats it accepts and I'm concerned this could take years to play out. 

I'm also frustrated that I can't _show_ you, my dear reader, HDR. I'd love to be able to show you before and afters. But again, software is in the way, and well, if you don't have an HDR display then it's all for nought. 

Perhaps I can rig up a camera to shoot my HDR screen... in HDR. Remember, movies have no issues with HDR. But man, how clunky would that be... I doubt screen recordings in HDR are anywhere even on Apple's roadmap right now. But perhaps this is the way, the only way. I'll consider it but it would be a major hack. 

/rant LOL

**Comment by cfc62** [comment]
HDR file size comparisons:

AVIF: 2-4 MB
JPG: 40-50 MB
JPG-XL: 50-60 MB

Quality is OK across all formats when viewed natively, but HDR isn't rendered correctly in Photos with JPG (HDR not shown) and it's just plain poor in Preview. 

Oh yeah the RAW file size is less than 20MB, so these JPG sizes are crazy. AVIF looks like the winner, but it's not fully supported by Photos. 

Also there's no HDR marker in Photos for these files, it would appear Apple is reserving that for it's own files form iPhones. 

Note, LrC does not export in HEIF/HIEC, so I can't try that.
